docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+    that can be cached in compiled form.
+*   **Expressions are serializable and composable.** These properties are useful
+    in implementing upstream filter propagation. For example, a Broker could
+    advertise a single CEL expression composed of the union of its Triggers'
+    filters, allowing upstream sources to filter events before publishing to the
+    Broker.
+
+CEL is
+[memory-safe, side-effect-free, and terminating](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview),
+making it safe and secure to embed in a multi-tenant dispatcher process.
+
+## Examples
+
+### Single attribute exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+```
+
+Specified with a CEL expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.type == "com.github.pull.create"
+```
+
+### Multiple attributes exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+      source: https://github.com/knative/eventing/pulls/123
+```
+
+Specified with a CEL expression:
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" &&
+        ce.source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Prefix match on source
+
+Cannot be specified with the SourceAndType field.
+
+CEL doesn't support prefix match directly, but prefix match can be accomplished
+with a regular expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.source.matches("^https://github.com")
+```
+
+### Complex boolean expression
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" ||
+        (ce.type == "com.github.issue.create" && ce.source.matches("proposals")
+
+```
+
+### Parsed extensions
+
+Cannot be specified with the SourceAndType field.
+
+_This example assumes the repository name is available as a CloudEvents
+extension with name `repository`._
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseExtensions: true
+      expression: >
+        ce.type == "com.github.pull.create" ||
+         (ce.type == "com.github.issue.create" &&
+          ce.repository == "proposals")
+
+```
+
+### Parsed data
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+_Until [google/cel-go#203](https://github.com/google/cel-go/issues/203) is
+resolved, dynamic integer fields in CEL must be compared as floats._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseData: true
+      expression: >
+        ce.type == "dev.knative.observation" &&
+        data.latency > 300.0
+```
+
+## Caveats
```

Comment:
Add the caveat that using an expression language limits the number of supported platforms to those with runtimes for that language. A simpler structured filter interface could be implemented more easily on unusual platforms (for better or worse). 

We can work around this with sidecars, but that might bring its own problems.

Reply:
Various security groups have been discussing how to combine implicit and explicit logic to offer a simple getting started experience that is easy to extend, test, and maintain. After years of discussion we've landed at schema + expression.

Bearing with me on the strawman names, the following filter format has some nice properties:

```yaml
filter:
   matchTypes: "com.github.pull.create"
   matchSource: "https://github.com/*"
   expression: ce.repository == "proposals"
```

* Easy to move the type and source filtering upstream.
* Data-driven fields can be mechanically transformed to an expression and combined.
* Nice blend of easy to use and custom logic when it adds value.

Under the covers it could be CEL doing the evaluation of the whole filter:

```javascript
(!has(filter.matchTypes) || match(ce.type, filter.matchTypes)) &&
(!has(filter.matchSource) || match(ce.source, filter.matchSource)) &&
<expression>
```

Using CEL in this way allows you to try many different evaluation schemas quickly and cheaply, so I think it's worth a hard look.

Reply:
I think your proposed interface is reasonable @TristonianJones, and we will probably end up with something like it, but I'd like to get user feedback on the expression language first so we know where it needs help. With that feedback we can be sure any structured interfaces we add are actually what users want.

I like introducing the expression language first because it immediately unblocks most (all?) filtering use cases. Users can start using Triggers right away and then decide what annoys them the most.

@duglin [proposed on Slack](https://knative.slack.com/archives/C9JP909F0/p1555519051075900) that we drop `SourceAndType`, leaving CEL as the only way to specify a filter. I don't think this is the best long-term solution (sorry @duglin) but I like it in the short-term because it theoretically helps us learn which filters are annoying to specify in a CEL-only Trigger. In practice, I think we're stuck with `SourceAndType` for now because we've already cut a release with it (at least until we can start using CRD versioning).